File: ciel/evolution_hydra/TorchRLBridge.js
/**
 * ═══════════════════════════════════════════════════════════════
 * TORCH-RL-BRIDGE — Pont TypeScript ↔ Python RL trainer
 * ═══════════════════════════════════════════════════════════════
 *
 * PHASE 5 : TitanRL n'est plus une somme pondérée figée.
 * Il appelle un VRAI trainer RL (Python + numpy) qui fait du
 * policy gradient.
 *
 * Architecture :
 *   TitanRL.learn(state, reward)
 *     → TorchRLBridge.trainStep(state, reward)
 *       → spawn('python3', ['rl_trainer.py'])
 *         ← JSON result {weights, bias, loss, iterations}
 *       → checkpoint persisté
 *
 * Fallback : si Python indisponible, policy gradient en pur TS
 * (régression linéaire bayésienne simplifiée). Pas idéal mais
 * le code ne crashe pas.
 *
 * Sécurité :
 *  - trainerPath whitelisté (chemin relatif au repo)
 *  - subprocess avec stdio piped
 *  - timeout (30s) pour éviter les hangs
 *  - checkpoint dans /tmp (pas dans le code source)
 */
import { spawn } from 'child_process';
import * as fs from 'fs';
import logger from '../utils/logger.js';
const DEFAULT_TRAINER = 'hydra_vampire_v4/rl_trainer.py';
const DEFAULT_CHECKPOINT = '/tmp/hydra_rl_checkpoint.json';
const DEFAULT_TIMEOUT = 30_000;
export class TorchRLBridge {
    trainerPath;
    checkpointPath;
    timeoutMs;
    pythonBin;
    forceFallback;
    callCount = 0;
    fallbackCount = 0;
    constructor(options = {}) {
        this.trainerPath = options.trainerPath ?? DEFAULT_TRAINER;
        this.checkpointPath = options.checkpointPath ?? DEFAULT_CHECKPOINT;
        this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT;
        this.pythonBin = options.pythonBin ?? 'python3';
        this.forceFallback = options.forceFallback ?? false;
    }
    /**
     * Appelle le trainer Python pour faire UN step de policy gradient.
     * Si Python indisponible, fallback en pur TS.
     */
    async trainStep(state, reward) {
        this.callCount++;
        if (this.forceFallback) {
            this.fallbackCount++;
            return this.fallbackTrainStep(state, reward);
        }
        try {
            return await this.runTrainer({ state, reward });
        }
        catch (err) {
            logger.warn(`[TorchRL] ⚠️ Python trainer failed (${err.message}), using TS fallback`);
            this.fallbackCount++;
            return this.fallbackTrainStep(state, reward);
        }
    }
    /**
     * Variante "mock" : pour les tests, on injecte un faux subprocess.
     * Le mock respecte la même interface (Promise<RLStepResult>).
     */
    async trainStepWithMock(state, reward, mockFn) {
        this.callCount++;
        try {
            return await mockFn({ state, reward });
        }
        catch (err) {
            return {
                status: 'error',
                error: err.message,
            };
        }
    }
    /**
     * Lit le checkpoint existant (s'il y en a un) et le retourne.
     */
    loadCheckpoint() {
        if (!fs.existsSync(this.checkpointPath)) {
            return null;
        }
        try {
            const raw = fs.readFileSync(this.checkpointPath, 'utf-8');
            const ckpt = JSON.parse(raw);
            return {
                state: ckpt.state ?? new Array(12).fill(0),
                weights: ckpt.weights ?? new Array(12).fill(new Array(12).fill(0)),
                bias: ckpt.bias ?? new Array(12).fill(0),
            };
        }
        catch (err) {
            logger.warn(`[TorchRL] Could not load checkpoint: ${err.message}`);
            return null;
        }
    }
    /**
     * Réinitialise le checkpoint (utile pour les tests ou un nouveau run).
     */
    resetCheckpoint() {
        if (fs.existsSync(this.checkpointPath)) {
            fs.unlinkSync(this.checkpointPath);
        }
    }
    // ──────────────────────────────────────────────────────────
    // PRIVATE
    // ──────────────────────────────────────────────────────────
    async runTrainer(input) {
        return new Promise((resolve, reject) => {
            // Vérifier que python3 et le trainer existent
            if (!fs.existsSync(this.trainerPath)) {
                reject(new Error(`Trainer not found: ${this.trainerPath}`));
                return;
            }
            const env = {
                ...process.env,
                HYDRA_RL_CHECKPOINT: this.checkpointPath,
            };
            const proc = spawn(this.pythonBin, [this.trainerPath], {
                stdio: ['pipe', 'pipe', 'pipe'],
                env,
            });
            let stdout = '';
            let stderr = '';
            let resolved = false;
            const timeoutHandle = setTimeout(() => {
                if (!resolved) {
                    resolved = true;
                    proc.kill('SIGKILL');
                    reject(new Error(`Trainer timeout after ${this.timeoutMs}ms`));
                }
            }, this.timeoutMs);
            proc.stdout.on('data', (d) => {
                stdout += d.toString();
            });
            proc.stderr.on('data', (d) => {
                stderr += d.toString();
            });
            proc.on('error', (err) => {
                if (!resolved) {
                    resolved = true;
                    clearTimeout(timeoutHandle);
                    reject(err);
                }
            });
            proc.on('close', (code) => {
                if (resolved)
                    return;
                resolved = true;
                clearTimeout(timeoutHandle);
                if (code !== 0) {
                    reject(new Error(`Trainer exited with code ${code}: ${stderr.slice(0, 200)}`));
                    return;
                }
                // La dernière ligne de stdout est le JSON
                const lines = stdout.trim().split('\n');
                const lastLine = lines[lines.length - 1];
                try {
                    const result = JSON.parse(lastLine);
                    resolve(result);
                }
                catch (err) {
                    reject(new Error(`Trainer output not JSON: ${lastLine.slice(0, 200)}`));
                }
            });
            // Envoyer l'input
            try {
                proc.stdin.write(JSON.stringify(input));
                proc.stdin.end();
            }
            catch (err) {
                if (!resolved) {
                    resolved = true;
                    clearTimeout(timeoutHandle);
                    reject(err);
                }
            }
        });
    }
    /**
     * Fallback pur TypeScript : policy gradient sur 12D.
     * Pas de NN, juste une régression linéaire bayésienne simplifiée.
     * Mieux que rien quand Python n'est pas dispo.
     */
    async fallbackTrainStep(state, reward) {
        const ckpt = this.loadCheckpoint();
        let W = ckpt?.weights ?? new Array(12).fill(null).map(() => new Array(12).fill(0));
        let b = ckpt?.bias ?? new Array(12).fill(0);
        // Forward
        const action = new Array(12).fill(0);
        for (let i = 0; i < 12; i++) {
            let s = b[i];
            for (let j = 0; j < 12; j++) {
                s += W[i][j] * state[j];
            }
            action[i] = s;
        }
        // Loss = -sum(reward * log_softmax(action))
        const max = Math.max(...action);
        const exps = action.map(a => Math.exp(a - max));
        const sumExps = exps.reduce((a, b) => a + b, 0);
        const logSumExp = max + Math.log(sumExps);
        let loss = 0;
        for (let i = 0; i < 12; i++) {
            loss -= reward[i] * (action[i] - logSumExp);
        }
        // Update : gradient descent on loss
        const lr = 0.01;
        for (let i = 0; i < 12; i++) {
            for (let j = 0; j < 12; j++) {
                W[i][j] -= lr * (-reward[i] * state[j]);
            }
            b[i] -= lr * (-reward[i]);
        }
        // Persister le checkpoint
        try {
            fs.writeFileSync(this.checkpointPath, JSON.stringify({ weights: W, bias: b, state, iterations: (ckpt?.bias ? 1 : 0) + 1 }));
        }
        catch (err) {
            logger.warn(`[TorchRL] Could not save checkpoint: ${err.message}`);
        }
        return {
            status: 'fallback',
            fallback: true,
            loss,
            iterations: 1,
            weights: W,
            bias: b,
            action,
            meanReward: reward.reduce((a, b) => a + b, 0) / reward.length,
        };
    }
    // ──────────────────────────────────────────────────────────
    // OBSERVABILITÉ
    // ──────────────────────────────────────────────────────────
    getStats() {
        return {
            callCount: this.callCount,
            fallbackCount: this.fallbackCount,
            fallbackRate: this.callCount > 0 ? this.fallbackCount / this.callCount : 0,
            trainerAvailable: fs.existsSync(this.trainerPath),
        };
    }
}

File: evolution/torch_rl_bridge.py
from __future__ import annotations

import json
import math
import os
import tempfile
import time
from dataclasses import dataclass, field
from typing import Any


@dataclass(slots=True)
class RLState:
    state: list[float] = field(default_factory=lambda: [0.0] * 12)
    weights: list[list[float]] = field(default_factory=lambda: [[0.0] * 12 for _ in range(12)])
    bias: list[float] = field(default_factory=lambda: [0.0] * 12)


@dataclass(slots=True)
class RLStepResult:
    status: str = "error"
    loss: float = 0.0
    iterations: int = 0
    weights: list[list[float]] | None = None
    bias: list[float] | None = None
    action: list[float] | None = None
    mean_reward: float = 0.0
    error: str = ""
    fallback: bool = False


DIM = 12


class TorchRLBridge:
    def __init__(
        self,
        checkpoint_path: str = "",
        timeout_ms: int = 30_000,
        force_fallback: bool = False,
    ) -> None:
        self._checkpoint_path = checkpoint_path or os.path.join(
            tempfile.gettempdir(), "hydra_rl_checkpoint.json"
        )
        self._timeout_ms = timeout_ms
        self._force_fallback = force_fallback
        self._call_count = 0
        self._fallback_count = 0

    async def train_step(self, state: list[float], reward: list[float]) -> RLStepResult:
        self._call_count += 1
        if self._force_fallback:
            self._fallback_count += 1
            return self._fallback_train_step(state, reward)
        return self._fallback_train_step(state, reward)

    def load_checkpoint(self) -> RLState | None:
        if not os.path.isfile(self._checkpoint_path):
            return None
        try:
            with open(self._checkpoint_path, encoding="utf-8") as f:
                ckpt = json.load(f)
            return RLState(
                state=ckpt.get("state", [0.0] * DIM),
                weights=ckpt.get("weights", [[0.0] * DIM for _ in range(DIM)]),
                bias=ckpt.get("bias", [0.0] * DIM),
            )
        except (OSError, json.JSONDecodeError):
            return None

    def reset_checkpoint(self) -> None:
        if os.path.isfile(self._checkpoint_path):
            os.unlink(self._checkpoint_path)

    def _fallback_train_step(self, state: list[float], reward: list[float]) -> RLStepResult:
        ckpt = self.load_checkpoint()
        W = ckpt.weights if ckpt is not None else [[0.0] * DIM for _ in range(DIM)]
        b = ckpt.bias if ckpt is not None else [0.0] * DIM
        action = [0.0] * DIM
        for i in range(DIM):
            s = b[i]
            for j in range(DIM):
                s += W[i][j] * state[j]
            action[i] = s
        max_a = max(action)
        exps = [math.exp(a - max_a) for a in action]
        sum_exps = sum(exps)
        log_sum_exp = max_a + math.log(sum_exps)
        loss = 0.0
        for i in range(DIM):
            loss -= reward[i] * (action[i] - log_sum_exp)
        lr = 0.01
        for i in range(DIM):
            for j in range(DIM):
                W[i][j] -= lr * (-reward[i] * state[j])
            b[i] -= lr * (-reward[i])
        try:
            with open(self._checkpoint_path, "w", encoding="utf-8") as f:
                json.dump({"weights": W, "bias": b, "state": state, "iterations": 1}, f)
        except OSError:
            pass
        return RLStepResult(
            status="fallback",
            fallback=True,
            loss=loss,
            iterations=1,
            weights=W,
            bias=b,
            action=action,
            mean_reward=sum(reward) / len(reward) if reward else 0.0,
        )

    def get_stats(self) -> dict[str, Any]:
        return {
            "call_count": self._call_count,
            "fallback_count": self._fallback_count,
            "fallback_rate": self._fallback_count / self._call_count if self._call_count > 0 else 0.0,
            "checkpoint_exists": os.path.isfile(self._checkpoint_path),
        }

    def process(self, input_data: Any) -> dict[str, Any]:
        if isinstance(input_data, dict):
            state = input_data.get("state", [0.0] * DIM)
            reward = input_data.get("reward", [0.0] * DIM)
            result = self._fallback_train_step(state, reward)
            return {
                "status": result.status,
                "loss": result.loss,
                "mean_reward": result.mean_reward,
                "fallback": result.fallback,
            }
        return self.get_stats()
